xds/internal/balancer/ringhash/ringhash.go

```diff
@@ -121,3 +153,276 @@ func (sc *subConn) queueConnect() {
+	b.resolverErr = nil
+	if b.updateAddresses(s.ResolverState.Addresses) {
+		// If addresses were updated, no matter whether it resulted in SubConn
+		// creation/deletion, or just weight update, we will need to regenerate
+		// the ring.
+		var err error
+		b.ring, err = newRing(b.subConns, b.config.MinRingSize, b.config.MaxRingSize)
+		if err != nil {
+			panic(err)
+		}
+		b.regeneratePicker()
+		b.cc.UpdateState(balancer.State{ConnectivityState: b.state, Picker: b.picker})
+	}
+
+	// If resolver state contains no addresses, return an error so ClientConn
+	// will trigger re-resolve. Also records this as an resolver error, so when
+	// the overall state turns transient failure, the error message will have
+	// the zero address information.
+	if len(s.ResolverState.Addresses) == 0 {
+		b.ResolverError(errors.New("produced zero addresses"))
+		return balancer.ErrBadResolverState
+	}
+	return nil
+}
+
+func (b *ringhashBalancer) ResolverError(err error) {
+	b.resolverErr = err
+	if len(b.subConns) == 0 {
+		b.state = connectivity.TransientFailure
+	}
+
+	if b.state != connectivity.TransientFailure {
+		// The picker will not change since the balancer does not currently
+		// report an error.
+		return
+	}
+	b.regeneratePicker()
+	b.cc.UpdateState(balancer.State{
+		ConnectivityState: b.state,
+		Picker:            b.picker,
+	})
+}
+
+// UpdateSubConnState updates the per-SubConn state stored in the ring, and also
+// the aggregated state.
+//
+// It triggers an update to cc when:
+// - the new state is TransientFailure, to update the error message
+//   - it's possible that this is a noop, but sending an extra update is easier
+//   than comparing errors
+// - the aggregated state is changed
+//   - the same picker will be sent again, but this update may trigger a re-pick
+//   for some RPCs.
+func (b *ringhashBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	s := state.ConnectivityState
+	b.logger.Infof("handle SubConn state change: %p, %v", sc, s)
+	scs, ok := b.scStates[sc]
+	if !ok {
+		b.logger.Infof("got state changes for an unknown SubConn: %p, %v", sc, s)
+		return
+	}
+	oldSCState := scs.effectiveState()
+	scs.setState(s)
+	newSCState := scs.effectiveState()
+
+	var sendUpdate bool
+	oldBalancerState := b.state
+	b.state = b.csEvltr.recordTransition(oldSCState, newSCState)
+	if oldBalancerState != b.state {
+		sendUpdate = true
+	}
+
+	switch s {
+	case connectivity.Idle:
+		// When the overall state is TransientFailure, this will never get picks
+		// if there's a lower priority. Need to keep the SubConns connecting so
+		// there's a chance it will recover.
+		if b.state == connectivity.TransientFailure {
+			scs.queueConnect()
+		}
+		// Resend the picker, there's no need to regenerate the picker because
+		// the ring didn't change.
+		sendUpdate = true
+	case connectivity.Connecting, connectivity.Ready:
+		// Resend the picker, there's no need to regenerate the picker because
```

Comment:
Why re-send the picker on `Connecting`? I understand `Ready`

Reply:
Deleted.

Reply:
Oh I see.  `Ready->!Ready` for just 1 subconn doesn't need a new picker, since the ring is in the picker and it will adjust accordingly.